File: python/character_match.py
import re
# Part 1
def is_character_match(s1, s2):
	str1 = re.sub(r'[^\w]', '', s1.lower())
	str2 = re.sub(r'[^\w]', '', s2.lower())
	char1 = {}
	for char in str1:
		char1[char] = char1[char] +1 if char in char1 else 1
	for char in str2:
		if char not in char1:
			return False
		char1[char] -= 1
		if char1[char] < 0:
			return False
	return True


# Part 2
def anagrams_for(word, list_of_words):
	ans = []
	for w in list_of_words:
		if is_character_match(word, w):
			ans.append(w)
	return ans

File: js/characterMatch.js
exports.isCharacterMatch = function(str1, str2) {
  str1 = str1.toLowerCase()
  str2 = str2.toLowerCase()
  exceptions = [" ", "-", ".", "_", ",", "!"]
  let str1obj = {}
  for (let i=0; i<str1.length; i++) {
    if (exceptions.includes(str1[i])){
      continue
    }
    str1[i] in str1obj ? str1obj[str1[i]] += 1 : str1obj[str1[i]] = 1
  }
  for (let i=0; i<str2.length; i++) {
    if (exceptions.includes(str2[i])){
      continue
    }
    if (!(str2[i] in str1obj)) {
      return false
    }
    str1obj[str2[i]] -= 1;
    if (str1obj[str2[i]] < 0) {
      return false
    }
  }
  return true
};

exports.anagramsFor = function(word, listOfWords) {
  ans = [];
  listOfWords.forEach((option) => {
    if (this.isCharacterMatch(word, option)) {
      ans.push(option)
    }
  })
  return ans
};

